File: scripts/runReformatTree.m
%% reformat the tree structures for use


numinstance = length(allSTree);

allSKids = cell(1,numinstance);
empty = [];
for instance = 1:numinstance
    if mod(instance,1000) == 0
        disp(['Sentence: ' num2str(instance)]);
    end
    % get embeddings
    n = length(allSTree{instance});
    
    cnt = 0;
    for j = 1:length(allSStr{instance})
        if ~isempty(allSStr{instance}{j})
            cnt = cnt+1;
        end
    end
    if cnt < 2 % words in sentence
        empty = [empty instance];
        continue
    end
    
    t = tree2();
    t.pp = zeros(1,n);
    t.pp(1:n) = allSTree{instance};
    mostkids = length(find(allSTree{instance}==mode(allSTree{instance}))); % largest number of kids one node has
    t.kids = zeros(mostkids,n);
    for i = 1:n
        tempkids = find(allSTree{instance}==i);
        t.kids(1:length(tempkids),i) = tempkids;
    end
    
    t.leafFeatures = zeros(1,n);
    leafs = find(allSNum{instance}>0);
    t.isLeafnode = zeros(1,2*n);
    t.isLeafnode(leafs) = 1;
    
    t.pos = allSPOS{instance};
    
    for i = 1:length(leafs)
        t.leafFeatures(leafs(i)) = allSNum{instance}(leafs(i));
    end
    
    % binarize
    [inc numnode newt] = reformatTree(1, t, n+1);
    
    opp = zeros(1,2*numnode-1);
    okids = zeros(2*numnode-1,2);
    opos = cell(2*numnode-1,1);
    
    % reorder for trainRAE
    [pp nnextleaf nnextnode nkids pos] = reorder(1, newt, 1, 2*numnode-1, opp, okids, opos);
    
    % Change Note 11/8/11 Brody: The line below was replaced. Done so that
    % allSNum would be padded with -1's for nonterminals. 
    %newnum = zeros(1,numnode);
    newnum = -1*ones(1,length(pp));
    newstr = cell(1,numnode);
    newstrElem = cell(1,numnode);
    next = 1;
    for i=1:length(allSNum{instance})
        if (allSNum{instance}(i) > 0)
            newnum(next) = allSNum{instance}(i);
            newstr(next) = allSStr{instance}(i);
            newstrElem(next) = allSOStr{instance}(i);
            next = next + 1;
        end
    end
    
    allSNum{instance} = newnum;
    allSStr{instance} = newstr;
    allSOStr{instance} = newstrElem;
    allSTree{instance} = pp;
    allSKids{instance} = nkids;
    allSPOS{instance} = pos;
    
end

File: scripts/runReformatTree.py
#%% reformat the tree structures for use

import numpy as np
import scipy.stats as stats
from scripts.tree2 import tree2
from scripts.reformatTree import reformatTree
from scripts.reorder import reorder

def runReformatTree(allSTree, allSStr):
    
    numinstance = len(allSTree);
    empty = [];
    for instance in range( 0,numinstance):
        if instance % 1000 == 0:
            print 'Sentence: ', instance
        
        # get embeddings
        n = len(allSTree[instance]);
        
        cnt = 0;
        for j in range( 0,len(allSStr[instance])):
            if not (len(allSStr[instance][j])):
                cnt = cnt+1;
            

        if cnt < 2:  # words in sentence:
            empty.append(instance)
            continue
        
        
        t = tree2();
        t.pp = np.zeros(n);
        t.pp[:n] = allSTree[instance]
        mostkids = len(np.where(allSTree[instance] == stats.mode(allSTree[instance]))) # largest number of kids one node has
        t.kids = np.zeros((mostkids,n))
        for i in range(n):
            tempkids = np.where(allSTree[instance] == i);
            t.kids[0:len(tempkids),i] = tempkids;
        
        
        # binarize
        [inc, numnode, newt] = reformatTree(1, t, n+1);
        
        opp = np.zeros(2*numnode-1);
        okids = np.zeros((2*numnode-1,2));
        opos = np.zeros(2*numnode-1);
        
        [pp, nnextleaf, nnextnode, nkids, pos] = reorder(1, newt, 1, 2*numnode-1, opp, okids, opos);
        

        newstr = range(1,numnode);
        nxt = 1;
        for i in range(len(allSTree[instance])):
            newstr[nxt] = allSStr[instance][i];                
            nxt = nxt + 1;
            
        allSStr[instance] = newstr;       
        allSTree[instance] = pp;
        
        
        
    